gapipy/resources/base.py

```diff
@@ -28,14 +29,19 @@ def __init__(self, data, stub=False, client=None):
     def options(cls, client):
         return APIRequestor(client, cls).options()
 
-    def fetch(self):
+    def fetch(self, old_behaviour=False):
         logger.info('Fetching %s/%s', self._resource_name, self.id)
 
+        if old_behaviour:
+            httperrors_mapped_to_none = HTTPERRORS_MAPPED_TO_NONE
+        else:
+            httperrors_mapped_to_none = None
+
         # Fetch the resource using the client bound on it, which handles cache get/set.
         resource_obj = getattr(self._client, self._resource_name).get(
             self.id,
             variation_id=getattr(self, 'variation_id', None),
-            httperrors_mapped_to_none=None)
+            httperrors_mapped_to_none=httperrors_mapped_to_none)
 
         self._fill_fields(resource_obj._raw_data)
```

Comment:
I reckon this line breaks if you got back a `None` from the `get` call into `resource_obj`.

---

Prior to #119 there was an `if resource_obj` guarding the call to `_fill_fields`: https://github.com/gadventures/gapipy/commit/020db0982d03e3b4aa15bcc9b56d09bfc294dc4a#diff-4a48200e3a123d70f98cee3446ff35e8130ae90a5792f283cfd006e7ce21a8ca

I'm repeating myself, but I don't really see how it helps if that behaviour comes back.

If that is restored, and we image that we're going to do a `self.booking = self.departure_service.booking.fetch()` and we suppose that that results in a 403, then what happens?

- the `get` hands back a `None`
- the `_fill_fields` does not happen
- we still have whatever stub data we had before attempting the `get`
- a reference to `self` is returned

Isn't that just the same as if we had done `self.booking = self.departure_service.booking`? We'd get the stub data that was present on the `departure_service` but none of the rest of its fields.

Reply:
Something about your comments @jonprindiville made me realize that there might be another approach to solving my original problem. 

With the original snippet,
```python
self.booking = self.departure_service.booking.fetch()
```

I had assumed that we needed the `fetch()` call to get a `booking(stub)` object. But if the original behaviour was to just return `self` ... why not just grab that when things fail? So I tried this,

```python
try:
    self.booking = self.departure_service.booking.fetch()
except IOError:  # we get an HTTPError from downstream requests library
    self.booking = self.departure_service.booking
```

Which, to my delight, passes the failing unit test! :D 

Am going to see if this work-around, works, for the other failing tests too and if so, will withdraw this PR. Expect an update from me within an hour or so.

Reply:
🤦 I spoke too soon. Between all my open tabs, I forgot which were test code and which was code-under-test. My idea of catching `IOError` exceptions would have been fine for test-code ... but some of it also appears inside code-under-test ; which is not a change I can do without risking other things. 

Back to the drawing board.

Reply:
> catching IOError exceptions would have been fine for test-code ... but some of it also appears inside code-under-test ; which is not a change I can do without risking other things

Hm.

The old behaviour, though, if it were happening in production would mean that:
- we tried to fetch a thing
- there was an HTTPError and in fact nothing got fetched
- maybe later we get an AttributeError when trying to access the data we thought that we fetched

I think, though, that if we want the option of having that "old behaviour" on an explicit fetch, then maybe we can add an `httperrors_mapped_to_none=` kwarg on the fetch method?

I think that that kind of change would still also require:

- re-introducing a guard on the `_fill_fields` call, and
- changes to the affected application/test code; from `stuff.fetch()` to `stuff.fetch(httperrors_mapped_to_none=THE_ONES_WE_DONT_WANT_TO_HEAR_ABOUT)`

---

I wonder if I'm missing something here, though, happy to schedule a call if you want to chat about it 🤷